### PR TITLE
Windows: Add support for missing callback types

### PR DESCRIPTION
**Pull request by dgmcdona**
This PR updates the windows.callbacks.Callbacks plugin to support callback types that were present in the original volatility framework but have not yet been added to volatility3. These callback types include:

- `IoRegisterShutdownNotification`
- `IoRegisterFsRegistrationChange`
- `GenericKernelCallback`
- `EventCategoryHardwareProfileChange`
- `EventCategoryDeviceInterfaceChange`
- `EventCategoryTargetDeviceChange`
- `DbgSetDebugPrintCallback`

This required updates to the callbacks JSON symbol files, the creation of a `_SHUTDOWN_PACKET` extension, and updates to the plugin itself. Because it introduces three new requirements (handles, driverirp, and poolscanner), I have incremented the major version number for the callbacks plugin. No other plugins depend on the callbacks plugin at this time, so it was not necessary to increase version numbers in other plugins.

### CODE REVIEW

File: volatility3/framework/plugins/windows/callbacks.py
Comment (ikelos) [inline volatility3/framework/plugins/windows/callbacks.py]: We prefer to import `objects` and then use `objects.Pointer` or `objects.StructType`.

File: volatility3/framework/plugins/windows/callbacks.py
Comment (ikelos) [inline volatility3/framework/plugins/windows/callbacks.py]: And this is exactly the reason why, ObjectInterface is actually defined in `volatility3.framework.interfaces.objects` but we (mistakenly) imported it directly rather than the objects module.  This could lead to someone importing `callbacks.ObjectInterface`...  Please try hard to only ever import the modules (and the correct modules).  I'll likely fix the `symbols.windows.extensions` module to fix this mistake...

File: volatility3/framework/plugins/windows/callbacks.py
Comment (ikelos) [inline volatility3/framework/plugins/windows/callbacks.py:31]: Bumping this is the right thing to do, even if other modules in the core set don't import it.  Someone else somewhere might, so if there's been a non-backwards compatible change, bumping the major number is exactly the right thing to do.  5:)

File: volatility3/framework/plugins/windows/callbacks.py
Comment (ikelos) [inline volatility3/framework/plugins/windows/callbacks.py]: All of these version requirements check the same thing?  Either the individual components need checking (if they have versions) or just do the check once.

File: volatility3/framework/plugins/windows/callbacks.py
Comment (ikelos) [inline volatility3/framework/plugins/windows/callbacks.py]: This is clever, but I think he already have tools for doing it?  Should be that [`memory_object.get_symbol_table_name()`](https://github.com/volatilityfoundation/volatility3/blob/develop/volatility3/framework/interfaces/objects.py#L149) should do it?  It will also handle exceptional cases where the type may not have a symbol table (rare these days, but I believe still technically possible).

File: volatility3/framework/plugins/windows/callbacks.py
Comment (ikelos) [inline volatility3/framework/plugins/windows/callbacks.py]: This *shouldn't* require dereferencing?  `DeviceObject.DriverObject` should work.

File: volatility3/framework/plugins/windows/callbacks.py
Comment (ikelos) [inline volatility3/framework/plugins/windows/callbacks.py:281]: Good!  Thanks for using the right return types!  5:)

File: volatility3/framework/plugins/windows/callbacks.py
Comment (ikelos) [inline volatility3/framework/plugins/windows/callbacks.py]: Again the dereferencing *should* be unnecessary, although here it seems to be being used to check validity (I believe pointer have an `is_valid` method which would avoid the need for a potentially costly exception setup...

File: volatility3/framework/plugins/windows/callbacks.py
Comment (ikelos) [inline volatility3/framework/plugins/windows/callbacks.py]: Was this just for testing or do you think the message has no value?  Happy to entertain the idea, I genuinely don't know how useful it is, but I prefer knowing when things have gone wrong than just silently skipping over them.  Perhaps moving if from `debug` to `constants.LOGLEVEL_VVVV` would be better?

File: volatility3/framework/plugins/windows/callbacks.py
Comment (ikelos) [inline volatility3/framework/plugins/windows/callbacks.py:386]: I've never really used `cast` before, but I'm happy to have more accurate typing information if it helps.  Thanks!  5:)

File: volatility3/framework/plugins/windows/callbacks.py
Comment (ikelos) [inline volatility3/framework/plugins/windows/callbacks.py]: Again, debugging information shouldn't affect normal usage, so I'd prefer it were there for developers to figure out when things go wrong.  If it's too verbose, that's fine and we can look for a better way to get the right information across, but if this was just a forgotten comment then it's be best to reinstate the message if possible (happy to change it to `constants.LOGLEVEL_VVVV` if that'd be better?

File: volatility3/framework/plugins/windows/callbacks.py
Comment (ikelos) [inline volatility3/framework/plugins/windows/callbacks.py]: I'm interested to know under what circumstances this wouldn't be seen as a string by the type checker?  I'm all for improving the typing, but I'd prefer it not get out of hand manually casting everything rather than getting the types right at the source?

**Review by ikelos — COMMENTED** [review]
Thanks, there's a lot to go through here but it generally looks pretty good.  It's quite a significant chunk of functionality though so I'd quite like to defer to @iMHLv2 's opinion to ensure it does what it's supposed and there's no gotchas that I've missed...

File: volatility3/framework/plugins/windows/callbacks.py
Comment (dgmcdona) [reply volatility3/framework/plugins/windows/callbacks.py]: Thanks, I think I forgot to update this when copy-pasting. There should really be three separate requirements - the `handles.Handles` plugin, the `poolscanner.PoolScanner` plugin, and the list of major functions `driverirp.MAJOR_FUNCTIONS`, the last of which I'm not sure how to define a requirement for, since it's not a class in another module, but just a list of strings. I'm going to define it as a `PluginRequirement` for now, but let me know if there's a better way.

File: volatility3/framework/plugins/windows/callbacks.py
Comment (dgmcdona) [reply volatility3/framework/plugins/windows/callbacks.py]: I was wondering if I was getting carried away with the type-casting - I was trying to make things as explicit as possible but this may have been overkill, so I'll roll that change back.

File: volatility3/framework/plugins/windows/callbacks.py
Comment (ikelos) [inline volatility3/framework/plugins/windows/callbacks.py]: Please try not to import classes directly, but only ever modules.  Importing `ObjectInterface` like this may allow someone to one day write `from callbacks import ObjectInterface` and it then won't be clear that it actually belongs to another module.

File: volatility3/framework/plugins/windows/callbacks.py
Comment (ikelos) [inline volatility3/framework/plugins/windows/callbacks.py]: I'm generally a fan of having fixed values pulled out and tagged with a comment or put in a variable, but it doesn't feel like these will get used anywhere else, so I'd be tempted to just embed them in the code below?  The code below already kinda hard codes the binary tag values so I'm not sure it's much different?

File: volatility3/framework/plugins/windows/callbacks.py
Comment (ikelos) [inline volatility3/framework/plugins/windows/callbacks.py:251]: What objects *don't* have an `is_valid` method?  Feels like this could be changed to just check `pointers`?

File: volatility3/framework/plugins/windows/callbacks.py
Comment (ikelos) [inline volatility3/framework/plugins/windows/callbacks.py]: Yeah, probably a little overboard with the casting...  5;)  I mean, it's good and I approve of well typed systems, but... a lot of these things should already be the type they're cast to, and hopefully mypy will one day be able to figure that out, but I'm not sure littering our code with hints will help all that much?

**Review by ikelos — COMMENTED** [review]
Yep, pretty happy with this, just a few more comments.  5:)

**Comment by ikelos** [comment]
@iMHLv2 could you give this a scan over to check there's nothing subtle I've missed please?

File: volatility3/framework/plugins/windows/callbacks.py
Comment (dgmcdona) [reply volatility3/framework/plugins/windows/callbacks.py:251]: The check fails for `StructType` objects if we remove the `hasattr` check. For instance: `AttributeError: StructType has no attribute: callbacks-x641!_NOTIFICATION_PACKET.is_valid`

There is an explicitly defined `is_valid` method in the callbacks extension added as part of this PR, but I don't think any other objects will have an `is_valid` method since they are just `StructType` objects defined in the callbacks symbol files.

File: volatility3/framework/plugins/windows/callbacks.py
Comment (ikelos) [inline volatility3/framework/plugins/windows/callbacks.py]: `layer_name` isn't part of the function signature for this call... or used within it anywhere?

File: volatility3/framework/plugins/windows/callbacks.py
Comment (ikelos) [inline volatility3/framework/plugins/windows/callbacks.py]: Could we use `f"Unexpected object type {type_name}"` for consistency and readability please?

File: volatility3/framework/plugins/windows/callbacks.py
Comment (ikelos) [inline volatility3/framework/plugins/windows/callbacks.py]: Happy with this, but feels like using the more generic `interfaces.renderers.BaseAbsentValue` might be more useful to avoid changes in the future, since it covers all of them?

File: volatility3/framework/symbols/windows/extensions/callbacks.py
Comment (ikelos) [inline volatility3/framework/symbols/windows/extensions/callbacks.py]: Again, dereference() *shouldn't* need to be explicitly specified.  I'd prefer not to muddy the code with them when they're unnecessary in case people start picking up bad habits and assuming they're necessary.  If it *is* necessary, I'd like to know why in case we can improve the framework...

**Review by ikelos — COMMENTED** [review]
Thanks very much for your recent changes!  I'm sorry it's taken so long to get reviewed, I'll let @iMHLv2 know that it's still awaiting his approval before it can go in.  From a code perspective though it looks ok to me...  5:)

File: volatility3/framework/plugins/windows/callbacks.py
Comment (dgmcdona) [reply volatility3/framework/plugins/windows/callbacks.py:281]: 😄

**Comment by dgmcdona** [comment]
I fixed an additional issue with `IoRegisterShutdownNotification` callbacks on x64 samples. The `_SHUTDOWN_NOTIFICATION` extension class should only be applied on x86 samples; see [here](https://github.com/volatilityfoundation/volatility3/compare/78db3d8457528fd6f98bee6c0c8df2c8948b9d7d..4d6f4bba0d5127f4a78074266091cf03b808b0f9#diff-c8fa55cd12e00fa088afa5f3be4853ee5f91345ab1803bf22b0e081e7a050b8b) and [here](https://github.com/volatilityfoundation/volatility3/compare/78db3d8457528fd6f98bee6c0c8df2c8948b9d7d..4d6f4bba0d5127f4a78074266091cf03b808b0f9#diff-cde9a57b579c0b95ba3254cad67158a93390e44fba2e6d0f6892496b558b1315)

**Review by iMHLv2 — APPROVED** [review]
@ikelos This looks good now. Thanks @dgmcdona!

**Pull request merged**
